crypto/rand/rand_unix.c

```diff
@@ -325,6 +334,52 @@ static struct random_device {
 } random_devices[OSSL_NELEM(random_device_paths)];
 static int keep_random_devices_open = 1;
 
+#if defined(__linux)
+static void *shm_addr;
+
+# if !defined(FIPS_MODE)
+static void cleanup_shm(void)
+{
+    shmdt(shm_addr);
+}
+# endif
+#endif
+
+/*
+ * Ensure that the system randomness source has been adequately seeded.
+ * This is done by having the first start of libcrypto, read one bytes from
+ * /dev/random.  Subsequent starts of the library and later reseedings do not
+ * need to do this.
+ */
+static void wait_random_seeded(void)
+{
+#if defined(__linux)
+    static const char random_path[] = "/dev/random";
+    static int seeded = 0;
+    int shm_id, fd;
+    char c;
+
+    if (!seeded) {
+        if ((shm_id = shmget(OSSL_SHM_SEEDED, 1, 0)) == -1) {
+            if ((fd = open(random_path, O_RDONLY)) != -1) {
+                if (read(fd, &c, 1) == 1)
+                    shm_id = shmget(OSSL_SHM_SEEDED, 1, IPC_CREAT);
+                close(fd);
+            }
+        }
+        if (shm_id != -1) {
+            seeded = 1;
+            shm_addr = shmat(shm_id, NULL, SHM_RDONLY);
+# ifndef FIPS_MODE
+            /* TODO 3.0: The FIPS provider doesn't have OPENSSL_atexit */
```

Comment:
FIPS mode isn't going to work on older systems without good entropy, so remove the "TODO" part of the comment.

Reply:
Why wouldn't it work?

Getting FIPS validation using /dev/urandom as the entropy source requires justification that it has been adequately seeded.  This would provide some justification.

Reply:
yes, indeed.

Reply:
We'll need a big cleanup of the `TODO 3.0` comments.  This included.
I've merged with this present so we don't forget.

Worst case would be leaving it here, which leaks one page of address space (not memory) per time the FIPS provider is loaded.